crates/pico/src/memo.rs

```diff
@@ -0,0 +1,188 @@
+            if let Some(node) = db.storage_mut().derived_nodes().get_mut(&node_id) {
+                if node.value != value {
+                    node.value = value;
+                    state = MemoState::Computed;
+                }
+                node.dependencies = dependencies;
+                node.time_calculated = time_calculated;
+                node.time_verified = current_epoch;
+            } else {
+                db.storage_mut().derived_nodes().insert(
+                    node_id,
+                    DerivedNode {
+                        time_verified: current_epoch,
+                        time_calculated,
+                        dependencies,
+                        inner_fn,
+                        value,
+                    },
+                );
+                state = MemoState::Computed;
+            }
+            time_calculated
+        } else {
+            let node = db
+                .storage_mut()
+                .derived_nodes()
+                .get_mut(&node_id)
+                .expect("node should exist. This is indicative of a bug in Pico.");
+            node.time_verified = current_epoch;
+            node.time_calculated
+        }
+    } else {
+        let (value, dependencies, time_calculated) =
+            call_inner_fn_and_collect_dependencies(db, node_id.param_id, inner_fn);
+        db.storage_mut().derived_nodes().insert(
+            node_id,
+            DerivedNode {
+                time_verified: current_epoch,
+                time_calculated,
+                dependencies,
+                inner_fn,
+                value,
+            },
+        );
+        state = MemoState::Computed;
+        time_calculated
+    };
+    register_dependency(
+        db,
+        NodeKind::Derived(node_id),
+        time_calculated,
+        current_epoch,
+    );
+    state
+}
+
+fn any_dependency_changed<Db: Database>(
+    db: &mut Db,
+    node_id: &DerivedNodeId,
+    current_epoch: Epoch,
+) -> bool {
+    let dependencies = db
+        .storage()
+        .derived_nodes()
+        .get(node_id)
+        .expect("node should exist. This is indicative of a bug in Pico.")
+        .dependencies
+        .iter()
+        .filter_map(|dep| {
+            if dep.time_verified_or_calculated != current_epoch {
+                Some((dep.node_to, dep.time_verified_or_calculated))
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    dependencies
+        .into_iter()
+        .any(|(dep_node, time_verified)| match dep_node {
+            NodeKind::Source(key) => source_node_changed(db, &key, time_verified),
+            NodeKind::Derived(dep_node_id) => derived_node_changed(db, dep_node_id),
+        })
+}
+
+fn source_node_changed<Db: Database>(db: &Db, key: &Key, time_verified: Epoch) -> bool {
+    match db.storage().source_nodes().get(key) {
+        Some(source) => source.time_calculated > time_verified,
```

Comment:
nit: I think it would be clearer for the signature to be 

```
fn source_node_changed_since<Db: Database>(db: &Db, key: &Key, since: Epoch) -> bool {
```